### DISCUSSION TITLE
posts/Go/datatype/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#354] [Announcements]
https://www.liwenzhou.com/posts/Go/02_datatype/ 

 李文周的Blog

**Comment by chengkanghua** [comment]
字符串的申请方式，  s1 := "adsf"  运行会报语法错误
var s1 = "asdf" 这样可以


```
PS C:\Users\VULCAN\go_learm\src\ch2\test> go version
go version go1.12.7 windows/amd64
PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
package main

import "fmt"

s1 := `绗竴琛?绗簩琛?绗笁琛?`

func main() {

        fmt.Println(s1)
}
PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
# command-line-arguments
.\test.go:5:1: syntax error: non-declaration statement outside function body
PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
package main

import "fmt"

var s1 = `绗竴琛?绗簩琛?绗笁琛?`

func main() {

        fmt.Println(s1)
}
PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
第一行
第二行
第三行
```

**Reply by LIUWENLIANG001** [reply]
变量如果放在函数外面，必须要加关键字var，如果在函数里面就可用可不用

**Comment by haidong5210** [comment]
> @chengkanghua
> 字符串的申请方式，  s1 := "adsf"  运行会报语法错误
> var s1 = "asdf" 这样可以
> 
> 
> ```
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go version
> go version go1.12.7 windows/amd64
> PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
> package main
> 
> import "fmt"
> 
> s1 := `绗竴琛?绗簩琛?绗笁琛?`
> 
> func main() {
> 
>         fmt.Println(s1)
> }
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
> # command-line-arguments
> .\test.go:5:1: syntax error: non-declaration statement outside function body
> PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
> package main
> 
> import "fmt"
> 
> var s1 = `绗竴琛?绗簩琛?绗笁琛?`
> 
> func main() {
> 
>         fmt.Println(s1)
> }
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
> 第一行
> 第二行
> 第三行
> ```

不能在funk 外面 ：=声明变量 在外面只能var

**Comment by Q1mi** [comment]
> @chengkanghua
> 字符串的申请方式，  s1 := "adsf"  运行会报语法错误
> var s1 = "asdf" 这样可以
> 
> 
> ```
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go version
> go version go1.12.7 windows/amd64
> PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
> package main
> 
> import "fmt"
> 
> s1 := `绗竴琛?绗簩琛?绗笁琛?`
> 
> func main() {
> 
>         fmt.Println(s1)
> }
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
> # command-line-arguments
> .\test.go:5:1: syntax error: non-declaration statement outside function body
> PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
> package main
> 
> import "fmt"
> 
> var s1 = `绗竴琛?绗簩琛?绗笁琛?`
> 
> func main() {
> 
>         fmt.Println(s1)
> }
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
> 第一行
> 第二行
> 第三行
> ```

Go语法规定程序中的代码必须都以关键字开头，s1 := `xxx` 是以变量开头的。

**Comment by asd6335683** [comment]
func main(){
	str:="hello沙河小王子"
	temp:=[]rune(str)
	var count int
	for _,v:=range temp {
		if v > 256 {
			count ++
			fmt.Println(string(v))
		}
	}
	fmt.Println(count)
}

**Comment by ghost** [comment]
```go
package main

import "fmt"

func main() {
	words := "hello沙河小王子"
	count := 0
	const startCode = 0x2E80
	const endCode = 0x9FFF
	for _, w := range words {
		if startCode < w && endCode > w {
			count++
		}
	}
	fmt.Println("中文字符数:", count)
}
```

**Comment by ghost** [comment]
代码缩进8个空格,接受不了///

**Comment by huzibbs** [comment]
```
package main

import "fmt"

func main() {
	var (
		str  = "hello沙河小王子"
		char = []rune(str)
		d    = 0
	)
	for i := 0; i < len(char); i++ {
		if len(string(char[i])) >= 3 {
			d++
		}
	}
	fmt.Printf("汉字总计数量：%d", d)
}
```

**Comment by Zhu-Ying-Zhang** [comment]
``` 
package main

import "fmt"

func main() {
        s := "hello 你好"
	count := 0
	for _, v := range s {
		fmt.Printf("%v\n", v)
		fmt.Println(len(string(v)))
		if len(string(v)) >= 3 {
			count++
		}
	}
	fmt.Println("漢字數量", count)
}

**Comment by zhonglanchengaijingyi** [comment]
package main

import (
	"fmt"
)

func main() {
	a := 1
	var b float64 = 1
	c := true
	d := "123"
	fmt.Printf("%d:%T\n", a, a)
	fmt.Printf("%v:%T\n", b, b)
	fmt.Printf("%v:%T\n", c, c)
	fmt.Printf("%s:%T\n", d, d)

	fmt.Println("-------------------")

	str := "hello沙河小王子"
	count := 0
	for _, s := range str {
		fmt.Printf("%c\n", s)
		if len(string(s)) >= 3 {
			count++
		}
	}
	fmt.Printf("\"%s\"中的汉子数量是:%d", str, count)

}

**Comment by shuwenjunn** [comment]
```
//乘法口诀表
package main

import "fmt"

func main() {
	for i := 1; i < 10; i++ {
		for j := 1; j < 10; j++ {
			// fmt.Println(i * j)
			if j <= i {
				fmt.Printf("%d*%d=%d ", i, j, i*j)
			}

		}
		fmt.Println("")
	}
}

```

**Comment by mytea003** [comment]
```go
func main() {
	//1、编写代码分别定义一个整型、浮点型、布尔型、字符串型变量，使用fmt.Printf()搭配%T分别打印出上述变量的值和类型。
	a, b, c, d := 100, 3.14, true, "hello"
	fmt.Printf("a的值是%d是%T类型--b的值是%f是%T类型--c的值是%t是%T类型--d的值是%s是%T类型\n", a, a, b, b, c, c, d, d)
	//2、编写代码统计出字符串"hello沙河小王子"中汉字的数量。
	var str string = "hello沙河小王子"
	count := 0
	for _, v := range str {
		if len(string(v)) >= 3 {
			count++
		}
	}
	fmt.Printf("汉字的数量为%d", count)
	str2 := "a"
	fmt.Println("单个字母的string类型长度为", len(str2))
	str2 = "人"
	fmt.Println("单个汉字的string类型长度为", len(str2))
}
```
第二题不会，看了各位同学的，再找了一下资料。bool，byte的长度均为1，rune的长度为4。 一开始卡在哪里呢，卡在怎么统计汉字，想来想去都想不出。看了别人的代码之后也不明白为什么是大于等于3，再看了资料才懂。一开始是想不到len()这个函数，统计这个长度应该怎么搞。 看了文章的输出，h是等于104的，我还以为统计的是这个长度。所以没想明白。我还以为hello，len()函数会读出h是第0位，值是h。 看来我的思维被定死在那里了。所以一直想不出来。

**Comment by wyhhh** [comment]
```go
package main

import (
	`fmt`
	`unicode/utf8`
)

func main() {
	s := `我ddd方`
	fmt.Println((len(s) - utf8.RuneCountInString(s)) / 2)
}
```
计算汉字数量 我们可以根据**字节数**与**字符数**差值进行计算(原理参见鸡兔同笼问题), 在u8编码下, 那么一个汉字字节数比一个汉字字符数多2, 那么有一汉字就多一个2 有两个汉字就多两个2 有n个汉字就多n个2 那么最后再除以2即可. 注意如果有非BMP平面字 也就是码点大于`\uffff`, 对应的u8就是4个字节了 比如`𪚥`字就占u8的4个字节 这时候你再用这个方法计算就可能不准确了(如果只有一个这样的字答案是准确的, 多余一个就不准确了 因为一个这样的字只多一个字节差 不到2 所以除以2直接截断了) 那么需要去读取二进制位计算了 , 鉴于 我还不会go语言的一些操作 就用C++写吧
```cpp
// 0 ~ 21
//U+ 0000 ~ U+ 007F: 0XXXXXXX 0 ~ 7
//U+ 0080 ~ U+ 07FF: 110XXXXX 10XXXXXX 8 ~ 11
//U+ 0800 ~ U+ FFFF: 1110XXXX 10XXXXXX 10XXXXXX 12 ~ 16
//U+10000 ~ U+10FFFF: 11110XXX 10XXXXXX 10XXXXXX 10XXXXXX 17 ~ 21
constexpr int getChineseCharsCountInString(const char* s) {
    int count = 0;
    while (*s) {
        if (*s >= 0) {
            ++s;
            continue;
        }
        ++count;
        if ((*s & 0b111'00000) == 0b110'00000) { // put 2 chars u8charptr first
            s += 2;
        }
        else if ((*s & 0b1111'0000) == 0b1110'0000) { // put 3 chars u8charptr
            s += 3;
        }
        else if ((*s & 0b11111'000) == 0b11110'000) { // put 4 chars u8charptr
            s += 4;
        }
    }
    return count;
}
```

**Comment by John520** [comment]
```go

package main

import (
	"fmt"
	"unicode"
)

func main() {
	s := "这是个一包含汉字和英文的字符串, This is an apple"
	hzc := 0

	for _, v := range s {
		if unicode.Is(unicode.Han, v) {
			hzc++
		}
	}

	fmt.Printf("该字符串中找到中文个数是：%v\n", hzc)

}
```

**Comment by SmeagolYe** [comment]
done

**Comment by 32328254** [comment]
//1.编写代码分别定义一个整型、浮点型、布尔型、字符串型变量，使用fmt.Printf()搭配%T分别打印出上述变量的值和类型。
```
package main
import (
	"fmt"
)
var (
  a int
  b float64
  c bool
  str string


)

func main(){
  a = 10
  b = 3.14
  c = true
  str = "this is string"
  fmt.Printf("type:%T,value:%d\n",a,a) //十进制
  fmt.Printf("type:%T,value:%b\n",a,a) //十进制
  fmt.Printf("type:%T,value:%o\n",a,a) //八进制
  fmt.Printf("type:%T,value:%x\n",a,a) //十六进制
  fmt.Printf("type:%T,value:%f\n",b,b) //浮点数
  fmt.Printf("type:%T,value:%v\n",c,c) //bool型
  fmt.Printf("type:%T,value:%s\n",str,str) //字符串
}
```

**Comment by 32328254** [comment]
//2.编写代码统计出字符串"hello沙河小王子"中汉字的数量。

```
package main
import (
	"fmt"
)

func main(){
 count := 0
 str := "hello沙河小王子"
 for _,i := range str{
	 //fmt.Printf("%c\n",i)
	 length := len(string(i)) //获取每个字符串的长度
	 if length >= 3{ //判断字符长度
		 count ++
	 }
 }
 fmt.Printf("chinese number is %d\n",count)
}
```

**Comment by auula** [comment]
```go
        str := "ABCD上海Google公司"
	count := 0
	for _, v := range str {
		if len(string(v)) == 3 {
			count++
		}
	}
	fmt.Println(count)
```

**Comment by cutety** [comment]
var s = "hello沙河小王子"
	for _, r := range s {
		if unicode.Is(unicode.Han, r) {
			count++
		}
	}
	fmt.Println(count)

**Comment by SarletSKY** [comment]
```
package main

import (
	"unicode"
	"fmt"
)

/**
	unicode 包提供了一些测试Unicode码点属性的数据和函数.
	Is 报告该符文是否在指定范围的表中。
	Han 汉字的表范围
 */
func main(){
	count := 0
	s := "hello赵薇あなた"
	for _,val := range s {
		if unicode.Is(unicode.Han,val) {
			count ++
		}
	}
	fmt.Println("s中中文有",count,"个")
}
```
###注意一下这里是要求中文，不能用判断字节

**Comment by yangfan16** [comment]
// 编写代码统计出字符串"hello沙河小王子"中汉字的数量

```
func CalStringLen(val string) (num int) {
	content := []rune(val)
	for index := range content {
		if unicode.Is(unicode.Han, content[index]) {
			num = num + 1
		}
	}
	return
}
```

**Comment by LnkSaber** [comment]
func isHand(r rune) bool {
	return unicode.Is(unicode.Han, r)
}

func homeWork() {
	var a int = 9527
	var b float64 = 3.1415926
	var c bool = true
	var d string = "最爱吃牛肉"

	fmt.Printf("%T  a \n", a)
	fmt.Printf("%T  b \n", b)
	fmt.Printf("%T  c \n", c)
	fmt.Printf("%T  d \n", d)

	var str string = "hello林Saber"
	var num int = 0
	for _, r := range str {
		if isHand(r) {
			num++
		}
	}

	fmt.Println(num)

}

**Comment by jueyingbingfeng** [comment]
func main() {
	a := "hello沙河小王子"
	println(flashBack(a))
	println(stringLen(a))
}

func flashBack(a string) string {
	byteA := []rune(a)
	b := make([]rune, len(a))
	for k, v := range byteA[0:] {
		n := len(a) - k - 1
		b[n] = v
	}
	return (string(b))
}

func stringLen(a string) int {
	num := 0
	for _, v := range a {
		if len(string(v)) > 2 {
			num++
		}
	}
	return num
}

**Comment by WC-DaMao** [comment]
// 编写代码统计出字符串"hello沙河小王子"中汉字的数量 中文编码范围19968-40869
func hanCount() {
	var str string
	str = "hello沙河小王子日の"
	sss := []rune(str)
	k := 0
	for _, value := range sss{
		w := int(value)
		if w >= 19968 && w <= 40869{
			k ++
		}
	}
	fmt.Println(k)

**Comment by xiaotiandada** [comment]
👍

**Comment by seven779779** [comment]
func main() {
	s1 := "hello沙河小王子"
	s2 := []rune(s1)

	i := 0
	for _, v := range s2 {
		if len(string(v)) == 3 {
			i++
		}
	}
	fmt.Println(i)
}

**Comment by xjb714** [comment]
```package main
import("fmt")
func main(){
        s:="hello我是xxx年龄20性别x"
        temp:=[]rune(s)
        fmt.Println((len(s)-len(temp))/2)
}

**Comment by brianchiang-tw** [comment]
hw 1<br>
编写代码分别定义一个整型、浮点型、布尔型、字符串型变量，<br>
使用fmt.Printf()搭配%T分别打印出上述变量的值和类型。

```
package main

import (
	"fmt"
)


func main() {

	var(
		integer int64 = 12345
		floating float64 = 3.14159
		flag bool = true
		str string = "Hello world with type and value"
	)

	
	fmt.Printf("%v is of %T \n", integer, integer )
	fmt.Printf("%v is of %T \n", floating, floating)
	fmt.Printf("%v is of %T \n", flag, flag)
	fmt.Printf("%v is of %T \n", str, str)


	return
}

```

Output:
12345 is of int64
3.14159 is of float64
true is of bool
Hello world with type and value is of string

**Comment by brianchiang-tw** [comment]
hw 2<br>

编写代码统计出字符串"hello沙河小王子"中汉字的数量

```
package main

import (
	"fmt"
	"unicode"
)

func isNonAscii(s rune) bool {

	// return true for non-ascii rune
	return s > unicode.MaxASCII
}

func countNonAscii(s string) int{

	// compute the number of non-ascii rune

	var counter int = 0

	for _, singleRune := range s{
		if isNonAscii( singleRune ){
			counter++
		}
	}
	return counter
}


func main() {

	fmt.Printf("%v \n", countNonAscii("hello wolrd"))
	fmt.Printf("%v \n", countNonAscii("hello沙河小王子"))

	return
}
```

Output:
0
5

**Comment by ptxu78** [comment]
小问题：

转义符	含义
\r	        回车符（返回行首）
\n	        换行符（直接跳到下一行的同列位置）

这里应该是写反了？

**Comment by windf17** [comment]
~~~go
package main

import "fmt"

func CountC(s string) (n int) {

	for _, i := range s {
		if i >= 0X4E00 && i <= 0x9FA5 {  // 汉字的范围是0X4E00到0x9FA5
			n++
		}
	}
	return
}
func main() {
	s1 := "中a国b人c民d万岁"
	s2 := "最炫民族风style"
	fmt.Println(CountC(s1))
	fmt.Println(CountC(s2))
}
~~~

**Comment by Lilymz** [comment]
func pratice()  {
	/**
	*	记录汉字长度 unicode汉字范围0x2E86-0x9FFF
	*  @DATE 2021/9/30/10:07
	*  @params [data]
	*  @return
	**/
	defer func(){
		str :="hello沙河小王子你好"
		minLength :=0x2E86
		maxLength :=0x9FFF
		var i = 0
		for _,r := range str {
			if  r>int32(minLength) && r<int32(maxLength){
				i++
			}
		}
		fmt.Printf("汉字得长度是：%d",i)
	}()
	/**
	*	编写代码分别定义一个整型、浮点型、布尔型、字符串型变量，使用fmt.Printf()搭配%T分别打印出上述变量的值和类型。
	*  @DATE 2021/9/29/14:35
	*  @params
	*  @return
	**/
	defer func() {
		var(
			integer int
			float float32
			ifBool bool
			str string
		)
		fmt.Printf("%T  ",integer)
		fmt.Printf("%T  ",float)
		fmt.Printf("%T  ",ifBool)
		fmt.Printf("%T\n",str)
	}()
}

**Comment by QuYingying0628** [comment]
//统计出字符串"hello沙河小王子"中汉字的数量
	str2 := "hello沙河小王子"
	var num = 0
	for _, item := range str2 {
		if item>255 {
			num++
		}
	}
	fmt.Println("汉字数量为：",num)

**Comment by Dionysusio** [comment]
请教下，关于interface{} 类型的 struct 字段入库，该怎么处理比较好呢？

**Comment by huzibbs** [comment]
您发给我的信件已经收到。-----------------------------------------------------------

**Comment by 17370839053** [comment]
package main

import "fmt"
func main()  {
	s := "Myname沙河小王子厉害的人"
	i:=0
	for _, r := range s { //rune
		if int(r)>int('z') {
			i++
		}
		
	}
	fmt.Printf("%v", i)
}

**Comment by xyloneX** [comment]
package main

import (
	"fmt"
	"regexp"
)

func main() {

	str := "hello沙河小王子にほんご"
	count := 0
	for _, i := range str {
		if regexp.MustCompile("[\u4E00-\u9FA5]").MatchString(string(i)) {
			count += 1
		}
	}
	fmt.Printf("一共有%d个中文", count)
}

**Comment by huzibbs** [comment]
您发给我的信件已经收到。-----------------------------------------------------------

**Comment by yx619** [comment]
func main() { 
	 str := "hello沙河小王"
	 runeStr := []rune(str)
	 var count = 0
	 for i := 0; i < len(runeStr); i++ {
		if (runeStr[i] < 'a' || runeStr[i]  > 'z') {
			count += 1
		}
	 }
	 fmt.Println(count)
}

**Comment by zhzy-to** [comment]
```go

package main

import (
	"fmt"
	"unicode"
)

func countChineseCharacters(s string) int {
	count := 0
	for _, r := range s { // 遍历字符串的每个字符
		if unicode.Is(unicode.Han, r) { // 判断字符是否为汉字
			count++
		}
	}
	return count
}

func main() {
	// 整型、浮点型、布尔型、字符串型变量，使用fmt.Printf()搭配%T分别打印
	var a int64
	a = 1

	b := 3.141592643

	c := true

	d := "Hello World"

	fmt.Printf("a = %T\n", a)
	fmt.Printf("b = %T\n", b)
	fmt.Printf("c = %T\n", c)
	fmt.Printf("d = %T\n", d)

	// 统计出字符串"hello沙河小王子"中汉字的数量
	var str string
	str = "hello沙河小王子"

	fmt.Printf("str = %T\n", str)

	// byteStr := []byte(str)
	// runeStr := []rune(str)
	// fmt.Println(len(byteStr))
	// fmt.Println(len(runeStr))

	count := 0

	for _, r := range str {
		fmt.Println(string(r))
		if len(string(r)) >= 3 {
			count++
		}
	}

	fmt.Printf("count = %d\n", count)

	// fmt.Println(countChineseCharacters(str))
}
```

**Comment by John520** [comment]
邮件已收到，谢谢。
 
华南理工大学2015级控制理论与控制工程           徐将将            tel：13560349491（629491）            地址：华工（五三

**Comment by huzibbs** [comment]
您发给我的信件已经收到。-----------------------------------------------------------

**Comment by huzibbs** [comment]
您发给我的信件已经收到。-----------------------------------------------------------

**Comment by John520** [comment]
邮件已收到，谢谢。
 
华南理工大学2015级控制理论与控制工程           徐将将            tel：13560349491（629491）            地址：华工（五三

**Comment by chinababy-zlj** [comment]
parckage main
 import "fmt"
 import "unicode"
func main(){
a := "hello沙河大帅比"
count := 0
for _, r := range a {
if unicode.Is(unicode.han r){
count++
}
}
fmt.println("汉字数量为：count")
}

**Comment by huzibbs** [comment]
您发给我的信件已经收到。-----------------------------------------------------------

**Comment by Johnnylov** [comment]
package main
import ("fmt")
func main() {
	var a int = 10
	var b float64 = 20.0
	var c bool = true
	var d string = "hello"
	fmt.Printf("%T %v\n",a,a)
	fmt.Printf("%T %f\n",b,b)
	fmt.Printf("%T %t\n",c,c)
	fmt.Printf("%T %s\n",d,d)
}
package main
import ("fmt")
func main() {
s := "hello沙河小王子"
var count int = 0
for _,r := range s {
	_ = r
	if r > 255 {
	count++
}
}
fmt.Println("字符个数为:",count)
}
